[Runtime] Init Runtime component (#15081)
symfony/symfony-docs: components/runtime.rst

.. index::
   single: Runtime
   single: Components; Runtime

The Runtime Component
======================

    The Runtime Component decouples the bootstrapping logic from any global state
    to make sure the application can run with runtimes like PHP-FPM, ReactPHP,

>     Swoole etc without any changes.

Reviewer: ```suggestion
    Swoole, etc. without any changes.
```

Installation
------------

.. code-block:: terminal

    $ composer require symfony/runtime

.. include:: /components/require_autoload.rst.inc

Usage
-----

The Runtime component allows you to write front-controllers in a generic way
and with use of configuration you may change the behavior. Let's consider the
``public/index.php`` as an example. It will return a callable which will create
and return the application::

    <?php
    // public/index.php
    use App\Kernel;

    require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

    return function (array $context) {
        return new Kernel($context['APP_ENV'], (bool) $context['APP_DEBUG']);
    };

The special ``autoload_runtime.php`` is automatically created when composer dumps
the autoload files since the component includes a composer plugin. The ``autoload_runtime.php``
will instantiate a :class:`Symfony\\Component\\Runtime\\RuntimeInterface`, its job
is to take the callable and resolve the arguments (``array $context``). Then it calls
the callable to get the application ``App\Kernel``. At last it will run the application,

> ie calling ``$kernel->handle(Request::createFromGlobals())->send()``.

Reviewer 2: ```suggestion
i.e. calling ``$kernel->handle(Request::createFromGlobals())->send()``.
```
maybe?

Author: Yes. Thank you

To make a console application, the same bootstrap code would look like::

    #!/usr/bin/env php
    <?php
    // bin/console

    use App\Kernel;
    use Symfony\Bundle\FrameworkBundle\Console\Application;

    require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

    return function (array $context) {
        $kernel = new Kernel($context['APP_ENV'], (bool) $context['APP_DEBUG']);
        return new Application($kernel);
    };

Selecting Runtimes
------------------

> The default Runtime is :class:`Symfony\\Component\\Runtime\\SymfonyRuntime`, it
> works excellent on most applications running with a webserver like Nginx and Apache,
> and PHP-FPM. You may change Runtime to :class:`Symfony\\Component\\Runtime\\GenericRuntime`

Reviewer: ```suggestion
The default Runtime is :class:`Symfony\\Component\\Runtime\\SymfonyRuntime`. It
works excellent on most applications running with a webserver using PHP-FPM like Nginx or Apache. You may change Runtime to :class:`Symfony\\Component\\Runtime\\GenericRuntime`
```

> or a custom Runtime for Swoole or Aws Lambda. This can be done by specifying the

Reviewer 3: ```suggestion
or a custom Runtime for Swoole or AWS Lambda. This can be done by specifying the
```
?

> Runtime class in the ``APP_RUNTIME`` environment variable or to specify the

Reviewer: ```suggestion
Runtime class in the ``APP_RUNTIME`` environment variable or by specifying the
```

``extra.runtime.class`` in ``composer.json``.

.. code-block:: json

    {
        "require": {
            "...": "..."
        },
        "extra": {
            "runtime": {
                "class": "Symfony\\Component\\Runtime\\GenericRuntime"
            }
        }
    }

> Using SymfonyRuntime
> --------------------

Reviewer 3: ```suggestion
Using the SymfonyRuntime
------------------------
```

The :class:`Symfony\\Component\\Runtime\\RuntimeInterface` has two methods. One
to get an instance of :class:`Symfony\\Component\\Runtime\\ResolverInterface`

> that prepares the arguments to the callable and one get an instance of

Reviewer: ```suggestion
that prepares the arguments to the callable and one to get an instance of
```

    // bin/console

    use App\Kernel;
    use Symfony\Bundle\FrameworkBundle\Console\Application;

    require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

    return function (array $context) {
        $kernel = new Kernel($context['APP_ENV'], (bool) $context['APP_DEBUG']);
        return new Application($kernel);
    };

Selecting Runtimes
------------------

The default Runtime is :class:`Symfony\\Component\\Runtime\\SymfonyRuntime`, it
works excellent on most applications running with a webserver like Nginx and Apache,
and PHP-FPM. You may change Runtime to :class:`Symfony\\Component\\Runtime\\GenericRuntime`
or a custom Runtime for Swoole or Aws Lambda. This can be done by specifying the
Runtime class in the ``APP_RUNTIME`` environment variable or to specify the
``extra.runtime.class`` in ``composer.json``.

.. code-block:: json

    {
        "require": {
            "...": "..."
        },
        "extra": {
            "runtime": {
                "class": "Symfony\\Component\\Runtime\\GenericRuntime"
            }
        }
    }

Using SymfonyRuntime
--------------------

The :class:`Symfony\\Component\\Runtime\\RuntimeInterface` has two methods. One

> to resolve get an instance of :class:`Symfony\\Component\\Runtime\\ResolverInterface`

Reviewer 4: You can remove `resolve` here.

Author: Thank you for proof reading.

``void``
    If the callable doesn't return anything, the ``SymfonyRuntime`` will assume
    everything is fine.::

        <?php

        require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

        return function () {
            echo 'Hello world';
        };

Using Options
~~~~~~~~~~~~~

The :class:`Symfony\\Component\\Runtime\\SymfonyRuntime` supports a number of
options. They can be passed to the constructor in two ways. First by specifying
options to ``APP_RUNTIME_OPTIONS`` environment variable.::

    <?php

    $_SERVER['APP_RUNTIME_OPTIONS'] = [
        'project_dir' => '/var/task',
    ];

    require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

    // ...

The second way to pass an option to ``SymfonyRuntime::__construct()`` is to use
``extra.runtime.options`` in ``composer.json``.

.. code-block:: json

    {
        "require": {
            "...": "..."
        },
        "extra": {
            "runtime": {
                "options": {
                    "project_dir": "/var/task"
                }
            }
        }
    }

.. note::

    The environment variable ``APP_DEBUG`` has special support to easily
    turn on and off debugging.

> Creating Your Own Runtime
> -------------------------

Reviewer 3: ```suggestion
Create Your Own Runtime
-----------------------
```

This is an advanced topic that describes the internals of the Runtime component.

Using the runtime component will benefit maintainers because the bootstrap logic
could be versioned as a part of a normal package. If the application author decides
to use this component, the package maintainer of the Runtime class will have more
control and can fix bugs and add features.

> -- note::

Reviewer 3: ```suggestion
.. note::
```

Author: oops. Thank you

            });

            $socket = new \React\Socket\Server($this->port, $loop);
            $server->listen($socket);

            $loop->run();

            return 0;
        }
    }

Now we should create a new :class:`Symfony\\Component\\Runtime\\RuntimeInterface`
that is using our ``ReactPHPRunner``::

    use Symfony\Component\Runtime\GenericRuntime;
    use Symfony\Component\Runtime\RunnerInterface;

    class ReactPHPRuntime extends GenericRuntime
    {
        private $port;

        public function __construct(array $options)
        {
            $this->port = $options['port'] ?? 8080;
            parent::__construct($options);
        }

        public function getRunner(?object $application): RunnerInterface
        {
            if ($application instanceof RequestHandlerInterface) {
                return new ReactPHPRunner($application, $this->port);
            }

            return parent::getRunner($application);
        }
    }

The end user will now be able to create front controller like::

    require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

    return function (array $context) {
        return new Psr15Application();
    };

> 

Reviewer 3: ```suggestion
```

    The Runtime Component decouples the bootstrapping logic from any global state
    to make sure the application can run with runtimes like PHP-FPM, ReactPHP,
    Swoole, etc. without any changes.

.. versionadded:: 5.3

    The Runtime component was introduced in Symfony 5.3.

Installation
------------

.. code-block:: terminal

    $ composer require symfony/runtime

.. include:: /components/require_autoload.rst.inc

Usage
-----

The Runtime component abstracts most bootstrapping logic as so-called
*runtimes*, allowing you to write front-controllers in a generic way.
For instance, the Runtime component allows Symfony's ``public/index.php``
to look like this::

    <?php
    // public/index.php
    use App\Kernel;

    require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

    return function (array $context) {
        return new Kernel($context['APP_ENV'], (bool) $context['APP_DEBUG']);
    };

So how does this front-controller work? At first, the special
``autoload_runtime.php`` is automatically created by the Composer plugin in
the component. This file runs the following logic:

#. It instantiates a :class:`Symfony\\Component\\Runtime\\RuntimeInterface`;

> #. The callable (returned in the file) is passed to the Runtime, whose job

Reviewer: ```suggestion
#. The callable (returned by ``public/index.php``) is passed to the Runtime, whose job
```

        // application instead of the HTTP Kernel
        return new Application($kernel);
    };

Selecting Runtimes
------------------

The default Runtime is :class:`Symfony\\Component\\Runtime\\SymfonyRuntime`. It
works excellent on most applications running with a webserver using PHP-FPM like
Nginx or Apache.

The component also provides a :class:`Symfony\\Component\\Runtime\\GenericRuntime`,
which uses PHP's ``$_SERVER``, ``$_POST``, ``$_GET``, ``$_FILES`` and
``$_SESSION`` superglobals. You may also use a custom Runtime (e.g. to
integrate with Swoole or AWS Lambda).

Use the ``APP_RUNTIME`` environment variable or by specifying the
``extra.runtime.class`` in ``composer.json`` to change the Runtime class:

.. configuration-block::

    .. code-block:: json

        {
            "require": {
                "...": "..."
            },
            "extra": {
                "runtime": {
                    "class": "Symfony\\Component\\Runtime\\GenericRuntime"
                }
            }
        }

    .. code-block:: env

        APP_RUNTIME="Symfony\\Component\\Runtime\\GenericRuntime"

Using the Runtime
-----------------

> A Runtime is resposible for passing arguments into the closure and run the

Reviewer: ```suggestion
A Runtime is responsible for passing arguments into the closure and run the
```

                $output->write('Hello World');
            });

            return $command;
        };

:class:`Symfony\\Component\\Console\\Application`
    Useful with console applications with more than one command. This will use the
    :class:`Symfony\\Component\\Runtime\\Runner\\Symfony\\ConsoleApplicationRunner``::

        <?php

        use Symfony\Component\Console\Application;
        use Symfony\Component\Console\Command\Command;
        use Symfony\Component\Console\Input\InputInterface;
        use Symfony\Component\Console\Output\OutputInterface;

        require_once dirname(__DIR__).'/vendor/autoload_runtime.php';

        return function (array $context) {
            $command = new Command('hello');
            $command->setCode(function (InputInterface $input, OutputInterface $output) {
                $output->write('Hello World');
            });

            $app = new Application();
            $app->add($command);
            $app->setDefaultCommand('hello', true);

            return $app;
        };

The ``GenericRuntime`` and ``SymfonyRuntime`` also support these generic
applications:

:class:`Symfony\\Component\\Runtime\\RunnerInterface`

>     The ``RuntimeInterface`` is a way to use a custom application with the

Reviewer: ```suggestion
    The ``RunnerInterface`` is a way to use a custom application with the
```

#. The main entry point returns a *callable* (the "app") that wraps the application;
#. The *app callable* is passed to ``RuntimeInterface::getResolver()``, which returns
   a :class:`Symfony\\Component\\Runtime\\ResolverInterface`. This resolver returns
   an array with the app callable (or something that decorates this callable) at
   index 0 and all its resolved arguments at index 1.
#. The *app callable* is invoked with its arguments, it will return an object that
   represents the application.
#. This *application object* is passed to ``RuntimeInterface::getRunner()``, which
   returns a :class:`Symfony\\Component\\Runtime\\RunnerInterface`: an instance
   that knows how to "run" the appliction object.
#. The ``RunnerInterface::run(object $application)`` is called and it returns the
   exit status code as `int`.
#. The PHP engine is exited with this status code.

When creating a new runtime, there are two things to consider: First, what arguments
will the end user use? Second, what will the user's application look like?

For instance, imagine you want to create a runtime for `ReactPHP`_:

**What arguments will the end user use?**

> For a generic ReactPHP application, there are no special arguments are

Reviewer: ```suggestion
For a generic ReactPHP application, no special arguments are
```
